### PR TITLE
Add Metadata EXIF Provider

### PR DESCRIPTION
**Pull request by DominicChm**
This PR adds a class which connects to the Amiga's gRPC framework and is able to place arbitrary state into a PNG's EXIF data. Currently, it encodes all data as a JSON object into the `Exif.Photo.UserComment` tag (which isn't proper). This will be changed to encode into [`MakerNote`](https://exiv2.org/makernote.html) as soon as I figure out how to access that IFD using `Pillow`.

Once that's done, I can also encode into the proper fields for GPS, if it'd be useful. 

TODO:
- Test usability in MATLAB ([iminfo](https://www.mathworks.com/help/matlab/ref/imfinfo.html))
- Encode proper EXIF fields

### CODE REVIEW

**Comment by DominicChm** [comment]
@genkikondo When you have a second, mind reviewing my code style?

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        
+        dataDict = {}
+        print(self._last_service_messages)
+        
+        if m := self.get_message("gps", "/pvt"):
+            print("Set GPS location information")
+            dataDict["gps"] = m
+        
+        # Put JSON in Usercomment
+        exif[0x9286] = str(dataDict)
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:62]: Can we use:
from PIL.ExifTags import TAGS

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        """Returns the latest message from the given path and service"""
+        if (s := self._last_service_messages.get(service)) is not None:
+            if (p := s.get(path)) is not None:
+                return p
+        return None
+        
+    def add_exif(self, file):
+        """Opens the passed file, sets exif metadata, and saves it https://exiv2.org/tags.html
+        
+        Note: EXIF seems to be a bit of a mess and python support isn't great, so fields are
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:48]: is it a mess even with a library like piexif? Regardless sounds like we'd want to save to usercomment anyway for arbitrary fields.

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    _config = None
+    
+    def __init__(self, config):
+        self._load_config(config)
+    
+    def _load_config(self, config):
+        """Loads a standard Farm-NG service configuration, passed as a JSON dict"""
+        self._config = json_format.ParseDict(config, EventServiceConfigList())
+
+    async def init_clients(self):
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:21]: async is not needed here - we're not awaiting anything inside this method

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+import asyncio
+import os
+from google.protobuf import json_format
+from farm_ng.core.event_service_pb2 import EventServiceConfigList
+from farm_ng.core.event_client import EventClient
+from PIL import Image
+from PIL.ExifTags import GPS
+    
+class MetadataProvider:
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:9]: I'd separate the MetadataProvider from the Amiga gRPC client. MetadataProvider should take in the gRPC client as a dependency, and call into it to retrieve GPS data. That way, we centralize all calls to Amiga gRPC, and MetadataProvider's only job is to write the EXIF

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                "port": 3001,
+                "host": "129.65.121.182",
+                "log_level": "INFO",
+                "subscriptions": [
+                    {
+                        "uri": {
+                            "path": "*",
+                            "query": "service_name=gps"
+                        },
+                        "every_n": 1
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:81]: does this mean receive every message? Or receive a message every second?

**Review by genkikondo — COMMENTED** [review]
Looks great! Only high level comment is that I'd break out the service subscription logic into a separate class

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py (hunk 5 above)
Comment (DominicChm) [reply runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:81]: I believe every message - definitely getting messages from this subscription at > 1hz

**Comment by DominicChm** [comment]
@genkikondo Implemented your suggestions - take a look! Also tested on farm-ng and can confirm EXIF is recorded and accessible later (using this lib or `exiftool`). By default, I've added `filter` and `gps` to be logged but this can be expanded to any amiga datastream later. Currently, resulting data (retrieved with `exiftool -b <image>.png`) looks like

```json
{
    "gps": {
        "/pvt": {
            "stamp": {
                "stamp": 3677.281854325,
                "clockName": "denim-dandelion/monotonic",
                "semantics": "driver/receive"
            },
            "longitude": -123.1231532,
            "latitude": 30.123145523,
            "altitude": 103.3,
            "headingMotion": 2.6777853675063163e-05,
            "headingAccuracy": 3.0146668e-05,
            "groundSpeed": 0.202,
            "speedAccuracy": 1.04,
            "velNorth": 0.113,
            "velEast": 0.168,
            "velDown": 0.076,
            "status": {
                "timeFullyResolved": true,
                "gnssFixOk": true
            },
            "gpsTime": {
                "stamp": 1708459963.6,
                "clockName": "gps/POSIX",
                "semantics": "device/sample"
            },
            "horizontalAccuracy": 8.134,
            "verticalAccuracy": 9.792,
            "positionMode": 3,
            "pDop": 0.0236,
            "height": 70.944
        }
    },
    "filter": {
        "/state": {
            "pose": {
                "aFromB": {
                    "rotation": {
                        "unitQuaternion": {
                            "imag": {
                                "z": -0.14908058633932805
                            },
                            "real": 0.9888250496304805
                        }
                    },
                    "translation": {
                        "x": 0.38609316173534686,
                        "y": 0.7086588687185137
                    }
                },
                "frameA": "world",
                "frameB": "robot",
                "tangentOfBInA": {
                    "linearVelocity": {
                        "x": 1.9264001076072097e-07
                    },
                    "angularVelocity": {}
                }
            },
            "uncertaintyDiagonal": {
                "data": [
                    0.05300949795934314,
                    0.010714493203120165,
                    0.06744473968517002,
                    0.08219946522399121,
                    0.0,
                    0.0,
                    0.0
                ]
            },
            "heading": -0.29927680657260325,
            "divergenceCriteria": [
                "STD_DEV_X_HIGH",
                "STD_DEV_W_HIGH"
            ]
        }
    }
}
```

File: runner_image_capture_amiga_kivy/libs/amiga_client/amiga_client.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import asyncio
+from google.protobuf import json_format
+from farm_ng.core.event_service_pb2 import EventServiceConfigList
+from farm_ng.core.event_client import EventClient
+
+class AmigaClient:
+    _service_message_cache = {}
+    _subscription_tasks = []
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/libs/amiga_client/amiga_client.py:8]: We should probably cancel these tasks upon app quit

File: runner_image_capture_amiga_kivy/src/main.py [hunk 7]
```diff
@@ -27,18 +28,55 @@
...
+from metadata_provider.metadata_provider import MetadataProvider
+from amiga_client.amiga_client import AmigaClient
 
 
 CONFIG_FILE = "~/.config/runner-image-capture/config.json"
 CONFIG_KEY_SAVE_DIR = "saveDir"
 CONFIG_KEY_FILE_PREFIX = "filePrefix"
 CONFIG_KEY_EXPOSURE_MS = "exposureMs"
 CONFIG_KEY_INTERVAL_S = "intervalS"
+CONFIG_KEY_METADATA_SERVICES = "meta_services"
```
Comment (genkikondo) [inline runner_image_capture_amiga_kivy/src/main.py:40]: nit: s/meta_services/metaServices for consistency

**Review by genkikondo — APPROVED** [review]
Looks great! Just some minor comments

File: runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        """Loads JSON-encoded metadata from MakerNote"""
+        im = Image.open(file)
+        exif = im.getexif()
+        
+        if not Base.MakerNote in exif:
+            return None
+        
+        return json.loads(exif[Base.MakerNote])
+
+    def add_exif(self, file):
```
[the remark this answers is not in the document]
Comment (genkikondo) [reply runner_image_capture_amiga_kivy/libs/metadata_provider/metadata_provider.py:29]: and probably should return the filename

File: runner_image_capture_amiga_kivy/libs/amiga_client/amiga_client.py (hunk 6 above)
Comment (DominicChm) [reply runner_image_capture_amiga_kivy/libs/amiga_client/amiga_client.py:8]: Good catch 👍

**Pull request merged**